onnx/shape_inference/implementation.cc

```diff
@@ -471,6 +487,71 @@ class ShapeInferenceImplBase {
     }
   }
 
+
+ void process(const NodeProto& n, std::unordered_map<std::string, const AttributeProto*> attr_map) {
+    NodeProto copy_n(n);
+    // Add attribute information into the temporary node
+    copy_n.clear_attribute();
+    for (const auto& attr : n.attribute()) {
+      if (attr.has_ref_attr_name()) {
+        if (attr_map.count(attr.ref_attr_name())) {
+          auto copy_attr = *attr_map[attr.ref_attr_name()];
+          copy_attr.set_name(attr.name());
+          copy_n.add_attribute()->CopyFrom(copy_attr);
+        }
+      } else {
+        copy_n.add_attribute()->CopyFrom(attr);
+      }
+    }
+    process(copy_n);
+  }
+
+  void process(const FunctionProto& func_proto, InferenceContext& ctx) {
+    // Get a temporary tensor-shape map
+    const auto num_func_inputs = func_proto.input_size();
+    std::vector<TypeProto> types_cache(num_func_inputs);
+    for (int i = 0; i < num_func_inputs; ++i) {
+      if (ctx.getInputType(i) == nullptr) {
+        fail_type_inference("Input ", i, " type is missing.");
+      }
+      types_cache[i] = *ctx.getInputType(i); // TODO: investigate whether we can remove cache
+      value_types_by_name[func_proto.input().Get(i)] = &types_cache[i];
+    }
+
+    // Create a temporary initializer value map
+    for (int i = 0; i < static_cast<int>(ctx.getNumInputs()) && i < num_func_inputs; ++i) {
+      const TypeProto* type = ctx.getInputType(i);
+      if (type->value_case() == TypeProto::kTensorType && ctx.getInputData(i) != nullptr) {
+        input_data_by_name[func_proto.input().Get(i)] = ctx.getInputData(i);
+      } else if (type->value_case() == TypeProto::kSparseTensorType && ctx.getInputSparseData(i) != nullptr) {
+        input_sparse_data_by_name[func_proto.input().Get(i)] = ctx.getInputSparseData(i);
+      }
+    }
+
+    std::unordered_map<std::string, const AttributeProto*> attr_map;
+    for (auto& attr : func_proto.attribute()) {
+      if (ctx.getAttribute(attr) != nullptr) {
+        attr_map[attr] = ctx.getAttribute(attr);
+      }
+    }
+    for (auto& n : func_proto.node()) {
+      process(n, attr_map);
+    }
+
+    for (int i = 0; i < func_proto.output_size(); ++i) {
+      const std::string& output_name = func_proto.output().Get(i);
+      // Skip if no type inferred for the tensor
+      auto iter = value_types_by_name.find(output_name);
+      if (iter != value_types_by_name.cend()) {
+        // Copy the type info to ctx
+        // to pass back to main graph
+        auto type_proto = ctx.getOutputType(i);
+        type_proto->CopyFrom(*(iter->second));
+      }
+    }
+
+ }
+
  public:
   ShapeInferenceImplBase(
       GraphProto* gp,
```

Comment:
tiny nit: shall we use `g_in` instead of `gp` to make the naming convention same as others?

Reply:
Done